runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -222,6 +230,47 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
     return Optional.<Boolean>absent();
   }
 
+  /**
+   * Check if all data watermark value for each step of the pipeline reach to max.
+   *
+   * <p>If throwing IOException while retrieving job metrics, ignore and return false.
+   */
+  boolean atMaxWatermark(DataflowPipelineJob job) {
+    JobMetrics metrics;
+    try {
+      metrics = options.getDataflowClient().projects().jobs()
+          .getMetrics(job.getProjectId(), job.getJobId()).execute();
+    } catch (IOException e) {
+      LOG.warn("Ignore the exception {} while retrieving job metrics", e);
+      return false;
+    }
+
+    if (metrics == null || metrics.getMetrics() == null) {
```

Comment:
use guard style statements instead of increasing levels of nesting here and below.

For example:
if (metrics == null || metrics.getMetrics() == null) {
  LOG.warn("Metrics not present for Dataflow job {}.", job.getJobId());
  return false;
}

boolean hasMaxWatermark = false;
...